src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/Semantic/Services/DefaultRazorSemanticTokensInfoService.cs

```diff
@@ -80,18 +80,264 @@ internal class DefaultRazorSemanticTokensInfoService : RazorSemanticTokensInfoSe
+
+            // Try to concurrently process requests if there are multiple
+            var (startTokens, endTokens) = await ComputeAndCacheTokensInParallelAsync(
+                textDocumentIdentifier, documentSnapshot, documentVersion, semanticVersion,
+                requestedRange, cachedRange, cancellationToken).ConfigureAwait(false);
+
+            var absoluteLine = 0;
+
+            // a) Before processing the cached portion of the range, we might have to process tokens for the
+            // start of the range -> start of the cached range (if applicable).
+            if (cachedRange.Start.Line != requestedRange.Start.Line)
+            {
+                if (startTokens is null)
+                {
+                    return null;
+                }
+
+                absoluteLine = AddStartingTokens(requestedRange, cachedRange, finalTokens, startTokens);
+            }
+
+            // b) Add the cached range tokens to the results.
+            absoluteLine = AddCachedTokens(cachedTokens, finalTokens, absoluteLine);
+
+            // c) Process the ending range tokens (if applicable).
+            if (requestedRange.End.Line != cachedRange.End.Line)
+            {
+                if (endTokens is null)
+                {
+                    return null;
+                }
+
+                AddEndingTokens(requestedRange, cachedRange, finalTokens, absoluteLine, endTokens);
+            }
+
+            // We should now have all the tokens for the requested range (a combination of computed results and cached results).
+            return new SemanticTokens { Data = finalTokens.ToImmutableArray() };
+
+            async Task<(SemanticTokens? StartTokens, SemanticTokens? EndTokens)> ComputeAndCacheTokensInParallelAsync(
+                TextDocumentIdentifier textDocumentIdentifier,
+                DocumentSnapshot documentSnapshot,
+                int documentVersion,
+                VersionStamp semanticVersion,
+                Range requestedRange,
+                Range cachedRange,
+                CancellationToken cancellationToken)
+            {
+                var startTokensTask = Task.FromResult<SemanticTokens?>(null);
+                if (cachedRange.Start.Line != requestedRange.Start.Line)
+                {
+                    var partialRange = new Range { Start = requestedRange.Start, End = cachedRange.Start };
+                    startTokensTask = ComputeAndCacheTokensAsync(
+                        textDocumentIdentifier, documentSnapshot, documentVersion, semanticVersion,
+                        partialRange, cancellationToken);
+                }
+
+                var endTokensTask = Task.FromResult<SemanticTokens?>(null);
+                if (requestedRange.End.Line != cachedRange.End.Line)
+                {
+                    var partialRange = new Range { Start = cachedRange.End, End = requestedRange.End };
+                    endTokensTask = ComputeAndCacheTokensAsync(
```

Comment:
Just to clarify, if the user is in the midst of typing the "double request" approach here wouldn't be called because the document versions would be different right?

Reply:
Yep, it shouldn't be since the cache would be invalidated due to the updated document version